### PR TITLE
Prefer embedding to raising when possible

### PR DESCRIPTION
**Pull request by panglesd**
This PR changes the behaviour of ppxlib in case of error.

As of now, ppxlib is raising located errors whenever it encounter an error: For instance, if the parsing of a payload fails, a located error is raised and nothing happen at all, including all other context-free rewriting.
This is obviously very bad for user experience with merlin.

This PR changes that to embed the error at the right place whenever it makes sense to do it.

This requires propagating the errors through a `result` type. In order to use the `Result` monadic let-bindings, a common error type is used: `Location.Error.t NonEmptyList.t`, even if there can only be one error (and not a list). If a function returns both values and errors, the return type is `value_type * Location.Error.t list`.
Public raising functions were kept for backward compatibility, and new `*_res` functions were created. To avoid duplication od code, the implementation of the raising functions is to call the non raising function, and raise in case of `Error` output.

In some case, I felt it still makes sense to raise (or we might have no choice), and I kept the exception. An example is when we have bad locations (for instance they overlap). 

I also added a few (non exhaustive but covering the most frequent cases I think) tests.

Compare (with `derive_accessors` and `get_env` from ppxlib's examples):
image
(all errors are reported properly, `a` is defined as it should be), with
image
(`a` is not defined, the error on `get_env` is `Uninterpreted extension 'get_env'` so it is a misleading error, only the first error is reported properly)

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -5,6 +5,8 @@ unreleased
 
 - Fix quoter and optimize identifier quoting (#327, @sim642)
 
+- Embed ppxlib’s error instead of raising, to keep the AST (#329, @panglesd)
```
Comment (pitag-ha) [inline CHANGES.md:8]: I'd say there are two levels on which this PR has impact on the user: one is that the ppx driver now returns a meaningful AST also when there's a ppxlib internal error. And the other one is that lots of functions in the ppxlib library now have a version returning a result instead of raising. Do you agree? If so, maybe you could even point that out by adding two different points to the changelog entry? Also, the one entry you currently have is about the driver part (the first point I've mentioned), so it might also be nice to make that a bit clearer; e.g. something along the lines
```suggestion
- Driver: In the case of ppxlib internal errors, embed those errors instead of raising to return a meaningful AST (#329, @panglesd)
```

**Review by pitag-ha — COMMENTED** [review]
Thanks for improving this and making ppxlib set a good presedent! I have two thoughts about the PR.

The first and more important one: given that this PR adds 1500 lines of code and removes 500, I think it's natural to ask some questions about possible consequences: currently `ppxlib` is quite stable and has relatively few known bugs. How likely do you think it is that this PR is introducing new bugs? And similarly: how likely do you think it is that this PR implies a loss in performance? This second question is also why I'm asking about the two objects you're newly introducing (see comments below). And then also: adding 1000 new lines in netto to the project without introducing new modules means some loss in readability, at least to a certain extent. Do you think that loss of readability is just distributed gradually along all the code that's changed? Or can you point to places where a lot of new code is introduced that makes the code overly complex now?

And the second thought is just a suggestion about the documentation in the functions that still raise: I think it would be nice for the user to make clear why for a lot of functions we now have two versions (one that raises and one that returns a result) and which one to prefer. E.g., we could add to the documentation of the raising functions something along the lines "This function is kept for backwards-compatibility. In new code, please use <version_returning_result>.". What do you think?

**Comment by panglesd** [comment]
Thanks for your comments!

## About the increase in lines of code

TL;DR:
- The changes are scattered and not really fit to be put in a module
- There are few changes but they are repeated a lot
- Factorizing would be a big and deep refactoring using GADT context, I could try if that is what we want

About the net increase of 1000 lines, I would say that they are mostly scattered in the file, and the reason is that the AST type is so big that it is hard to make small changes, as any pattern-matching on the AST generates lots of cases, and small changes multiplied a lot is not small anymore.

Take for example this change: https://github.com/ocaml-ppx/ppxlib/pull/329/files#diff-6e863517f7732e2f783ffe103a3635a6576965e2afaa5af3f331440e50a10a93R762-R777

Dealing with the errors consists in getting the list of errors, turning them into the appropriate extension nodes, and prepending that list to the result. The logic is not that difficult, and its easy to review (for one of the function), even though it deletes 2 lines and add 12, a net of +10 loc.
However, the function `expand_sig_module_type_decl` has brothers and sisters `expand_str_type_decls`, `expand_sig_type_decls`, `expand_str_module_type_decl`, `expand_sig_module_type_decl`, `expand_str_exception`, `expand_sig_exception`, `expand_str_type_ext`, `expand_sig_type_ext`.
So that's approximately a +80 lines of code for the family, for basically +10 same lines of code repeated 8 times. And the same phenomenon happen in many other places, especially in `context_free.ml` (basically any place that match on the ast node and does not use a GADT context trick).

The reason all of this is not factorized is, I think, the following: those function act on different types, so you'll have to use a GADT trick to factorize them (as it is done in some places), and moreover subtle changes in the way we treat various nodes of the AST might make it hard to factorize.

I could try to factorize more, but in the end, it will also be a big refactoring with many lines changed (with a big net loss this time), and potential errors.

About your question whether the PR is likely to contain errors: How dare you say there is a chance it contains a single error? More seriously, it is a very difficult question to answer! I would say that it is mostly safe, as it really is a "type-driven" modification: `grep` the source to find occurrences of `raise_errorf`, transform the return type of the function into a `result`, propagate using monad style, and embed in the end.
So in the end, I will say "not super likely", but I will wrongly invoke Rice's theorem to argue that I cannot be sure that I did not introduce bugs.

I would like to mention another option: instead of propagating the `result` type, we could have kept the raising functions and catch them at the points were we want to embed them. This changes the code way less, but is much less good for several reasons:
- Sometimes, such as when traversing the AST looking for unresolved attributes, we would raise and return only one error when we could have returned many
- It does not add the functions for the library user to have results instead of exceptions

So I would be against this, you can try to convince me though...

## About the loss of performance

I don't think it implies a big loss of performance, especially when there is no error.

When there are error, ppxlib used to stop at the first one. Now, it continues to collect all of them, so yes, it will take more time, but it will give all errors instead of just the first one.

## About the docs

Yes, it's a good idea to do that, I'll try to find the time to do it somehow!

File: src/common.ml [hunk 2]
```diff
@@ -157,14 +174,25 @@ let curry_applications expr =
...
+        None
+    | attr ->
+        let loc = loc_of_attribute attr in
+        Some (Location.Error.createf ~loc "Attributes not allowed here"))
+
+let no_attributes_errors_in =
+  object
+    inherit [Location.Error.t list] Ast_traverse.fold
+    method! attribute a acc = no_attributes_errors [ a ] @ acc
+  end
```
[the remark this answers is not in the document]
Comment (panglesd) [reply src/common.ml:190]: At exactly the same places as the corresponding raising object (`assert_no_attribute_in`): nowhere!
I guess it is for library user...
The name is a badly chosen though, I will change it.

File: src/attribute.ml [hunk 3]
```diff
@@ -394,44 +423,244 @@ module Floating = struct
...
-      "Attribute `%s' was not used" name
+    [
+      Name.Registrar.Error.createf registrar context ~white_list
+        "Attribute `%s' was not used" name;
+    ]
+  else []
+
+let collect_unused_attributes_errors =
+  object (self)
+    inherit [Location.Error.t list] Ast_traverse.fold as super
```
[the remark this answers is not in the document]
Comment (panglesd) [reply src/attribute.ml:468]: The reason we use an object is that it allows to inheritate from `Ast_traverse.fold`, which is a "base" for folding function, and just re-implements the parts that we need.

Before, we were using `check_attribute`, which would raise at the first unused attribute, so it is also a full traversal when there is no errors.

**Comment by pitag-ha** [comment]
Thanks, this looks all good to me! I don't think my doubts were meaningful, since no extra AST traversal is added.
